### ISSUE TITLE
Differences between the assert_eq macro and direct == operation

### ISSUE THREAD

**Issue by renyahua** [rust-lang/rust#158907]
I tried this code:

```rust
<code>
```

I expected to see this happen: *explanation*

Instead, this happened: *explanation*

### Meta

`rustc --version --verbose`:
```
<version>
```

<details><summary>Backtrace</summary>
<p>

```
<backtrace>
```

</p>
</details>

**Comment by saethlin** [comment]
@renyahua can you provide a reproducer? Or any explanation?

**Comment by jieyouxu** [comment]
Closing as missing info. Please feel free to reopen if there's more info for this report.

**Comment by renyahua** [comment]
```rust
fn main() {
    let (x, mut y) = (1_i32, 2_i32);
    let rx = &x;
    let ry = &mut y;

    // The first version compiles correctly.
    if !((&1, &2) == (rx, ry)) {
        panic!("not equal");
    }
    // The second version does not compile.
    assert_eq!((&1, &2), (rx, ry), "not equal");
}
```

**Comment by chenyukang** [comment]
For the first version:
```rust
    if !((&1, &2) == (rx, ry)) {
        panic!("not equal");
    }
```
the right-hand tuple is checked with the left-hand side as context. Since the left side has type roughly `(&i32, &i32)`, `ry: &mut i32` can be coerced/reborrowed as `&i32` while building the right-hand tuple.

`assert_eq!` is different, it expands roughly by first borrowing both arguments separately, so it can print them if the assertion fails:

https://github.com/rust-lang/rust/blob/bcd9ae8b77e8a1d03d73b51f1fd9cb99565cfdc5/library/core/src/macros/mod.rs#L42-L55

That means `(rx, ry)` is type-checked independently before it is compared with the left side. Without the expected type from `(&1, &2)`, its type remains `(&i32, &mut i32)`. 

The comparison then becomes:
```rust
(&i32, &i32) == (&i32, &mut i32)
```
which fails because the tuple element types differ in mutability.

if you force the coercion explicitly:
```rust
assert_eq!((&1, &2), (rx, &*ry), "not equal");
```
or give the right-hand side an explicit type:
```rust
let rhs: (&i32, &i32) = (rx, ry);
assert_eq!((&1, &2), rhs, "not equal");
```

the result will be equal.
